runtime/compiler/optimizer/OSRGuardInsertion.cpp

```diff
@@ -30,63 +30,205 @@
 #include "optimizer/Optimization_inlines.hpp"
 #include "optimizer/FearPointAnalysis.hpp"
 #include "optimizer/RematTools.hpp"
-#include "optimizer/HCRGuardAnalysis.hpp"
 #include "optimizer/TransformUtil.hpp"
 #include "ras/DebugCounter.hpp"
+#include "infra/Checklist.hpp"             // for TR::NodeChecklist
+
+TR_Structure* fakeRegion(TR::Compilation *comp);
 
 static bool generatesFear(TR::Compilation *comp, TR_FearPointAnalysis &fearAnalysis, TR::Block *block)
    {
    return fearAnalysis._blockAnalysisInfo[block->getNumber()]->get(0);
    }
 
-int32_t TR_OSRGuardInsertion::perform()
+
+static bool hasHCRGuard(TR::Compilation *comp)
    {
-   static char *disableOSRGuards = feGetEnv("TR_DisableOSRGuards");
+   TR::list<TR_VirtualGuard*> &virtualGuards = comp->getVirtualGuards();
+   auto guard = virtualGuards.begin();
+   for (; guard != virtualGuards.end(); ++guard)
+      {
+      if ((*guard)->getKind() == TR_HCRGuard || (*guard)->mergedWithHCRGuard())
+         return true;
+      }
 
-   // Currently, OSR guard insertion is only needed when in the OSR HCR mode
-   if (disableOSRGuards || !comp()->isOSRTransitionTarget(TR::postExecutionOSR))
+   return true;
+   }
+
+static bool hasYieldPointCannotOSR(TR::Compilation *comp)
+   {
+   for (TR::TreeTop *treeTop = comp->getStartTree(); treeTop; treeTop = treeTop->getNextTreeTop())
       {
-      if (trace())
-         traceMsg(comp(), "HCR Guards are only removed when in OSR HCR mode - skipping\n");
-      return 0;
+      TR::Node *ttNode = treeTop->getNode();
+
+      if (comp->isPotentialOSRPoint(ttNode) &&
+          !comp->isPotentialOSRPointWithSupport(treeTop))
+         {
+         traceMsg(comp, "Found first bad yield point at n%dn\n", ttNode->getGlobalIndex());
+         return true;
+         }
       }
 
-   // Even under NextGenHCR, OSR may have been disabled for this compilation at runtime
-   if (!comp()->supportsInduceOSR())
+   return false;
+   }
+
+void TR_OSRGuardInsertion::cleanUpPotentialOSRPointHelperCalls()
+   {
+   for (TR::TreeTop *treeTop = comp()->getStartTree(); treeTop; treeTop = treeTop->getNextTreeTop())
       {
-      if (trace())
-         traceMsg(comp(), "OSR induce is not supported during this compilation - skipping\n");
-      return 0;
+      TR::Node *ttNode = treeTop->getNode();
+      if (ttNode->getNumChildren() != 1)
+         continue;
+
+      TR::Node *node = ttNode->getFirstChild();
+      if (node->isPotentialOSRPointHelperCall())
+         {
+         TR::TreeTop* prevTreeTop = treeTop->getPrevTreeTop();
+         TR::TransformUtil::removeTree(comp(), treeTop);
+         treeTop = prevTreeTop;
+         }
       }
+   }
 
-   // Detect if there are no HCR guards
-   TR::list<TR_VirtualGuard*> &virtualGuards = comp()->getVirtualGuards();
-   auto guard = virtualGuards.begin();
-   for (; guard != virtualGuards.end(); ++guard)
+void TR_OSRGuardInsertion::removeRedundantPotentialOSRPointHelperCalls(TR_HCRGuardAnalysis* guardAnalysis)
```

Comment:
Please provide a comment with an intuition for how this is working - I'm still working through the code to understand its interplay with the analyses in this opt pass.